solidity/random-beacon/contracts/ReimbursmentPool.sol

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: MIT
+//
+// ▓▓▌ ▓▓ ▐▓▓ ▓▓▓▓▓▓▓▓▓▓▌▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▄
+// ▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▌▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+//   ▓▓▓▓▓▓    ▓▓▓▓▓▓▓▀    ▐▓▓▓▓▓▓    ▐▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▄▄▓▓▓▓▓▓▓▀      ▐▓▓▓▓▓▓▄▄▄▄         ▓▓▓▓▓▓▄▄▄▄         ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▓▓▓▓▓▓▓▀        ▐▓▓▓▓▓▓▓▓▓▓         ▓▓▓▓▓▓▓▓▓▓         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▀▀▓▓▓▓▓▓▄       ▐▓▓▓▓▓▓▀▀▀▀         ▓▓▓▓▓▓▀▀▀▀         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▀
+//   ▓▓▓▓▓▓   ▀▓▓▓▓▓▓▄     ▐▓▓▓▓▓▓     ▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌
+// ▓▓▓▓▓▓▓▓▓▓ █▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+// ▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+//
+//                           Trust math, not hardware.
+
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+contract ReimbursementPool is Ownable, ReentrancyGuard {
+    /// @notice Authorized contracts that can interact with the reimbursment pool.
+    ///         Authorization can be granted and removed by the governance.
+    mapping(address => bool) public isAuthorized;
+
+    /// @notice Static gas of submitting a transaction.
+    uint256 public staticGas;
+
+    /// @notice Max gas price used to reimburse a transaction submitter. Protects
+    ///         against malicious operator-miners.
+    uint256 public maxGasPrice;
+
+    event StaticGasUpdated(uint256 staticGas);
+
+    event MaxGasPriceUpdated(uint256 maxGasPrice);
+
+    /// @notice Refunds ETH to a spender for executing specific transactions.
+    /// @dev Only authorized contracts are allowed calling this function.
+    /// @param gasSpent Gas spent on a transaction that needs to be reimbursed.
+    /// @param receiver Address where the reimbursment is sent.
+    function refund(uint256 gasSpent, address receiver) external nonReentrant {
+        require(
+            isAuthorized[msg.sender],
+            "Contract is not authorized for a refund"
+        );
+        require(receiver != address(0), "Receiver's address cannot be zero");
+
+        uint256 gasPrice = tx.gasprice < maxGasPrice
+            ? tx.gasprice
+            : maxGasPrice;
+
+        uint256 refundAmount = (gasSpent + staticGas) * gasPrice;
+
+        /* solhint-disable avoid-low-level-calls */
+        // slither-disable-next-line low-level-calls
+        (bool sent, ) = receiver.call{value: refundAmount}("");
+        /* solhint-enable avoid-low-level-calls */
+        require(sent, "Failed to refund Ether");
```

Comment:
We need to remember to catch this error when we use this function, so we don't end up with a failing transaction. Maybe it would be worth adding a note about it in `@dev` docs?

Reply:
I would ignore the result of `.call` in this function. For externally owned accounts `call` should always work. If the caller is a smart contract, it's that smart contract's responsibility to make sure Ether can be received. If there is some problem with receiving Ether, we do not need to revert, especially since we would revert at the very end, when the most complex and expensive logic has been already executed and paid for before `refund` is called. If we go this way, let's make sure to explain it in `@dev` section of docs.

Reply:
I'm on the fence between ignoring and wrapping a call in a try-catch on the caller side. The best practice is to have a `require` and if something goes south then a try-catch will inform about the error. I think this approach sort of "decouples" this contract and doesn't care who's going to call it. On the other hand, try-catch will add extra costs and at least for now, we know what contracts will call the `ReimbursmentPool` contract. Anyway, I removed the result of the `.call`, but also open for alternatives. [063786d](https://github.com/keep-network/keep-core/pull/2826/commits/063786d85371ffa16d1c170106990fd94acf1061)